engines/prover.cpp

```diff
@@ -377,38 +381,57 @@ bool Prover::witness(std::vector<UnorderedTermMap> & out)
     };
   }
 
-  for (int i = 0; i <= reached_k_; ++i) {
+  bool success = true;
+
+  // Some backends don't support full witnesses
+  // it will still populate state variables, but will return false instead of
+  // true
+  for (auto wit_map : witness_) {
     out.push_back(UnorderedTermMap());
     UnorderedTermMap & map = out.back();
 
     for (auto v : orig_ts_.statevars()) {
       SortKind sk = v->get_sort()->get_sort_kind();
-      Term vi = unroller_.at_time(transfer_to_prover_as(v, sk), i);
-      Term r = solver_->get_value(vi);
-      map[v] = transfer_to_orig_ts_as(r, sk);
+      Term pv = transfer_to_prover_as(v, sk);
+      map[v] = transfer_to_orig_ts_as(wit_map.at(pv), sk);
     }
 
     for (auto v : orig_ts_.inputvars()) {
       SortKind sk = v->get_sort()->get_sort_kind();
-      Term vi = unroller_.at_time(transfer_to_prover_as(v, sk), i);
-      Term r = solver_->get_value(vi);
-      map[v] = transfer_to_orig_ts_as(r, sk);
+      Term pv = transfer_to_prover_as(v, sk);
+      try {
+        map[v] = transfer_to_orig_ts_as(wit_map.at(pv), sk);
+      }
+      catch (std::exception & e) {
+        success = false;
```

Comment:
if `success` is false, we don't need to run the next loop.

Reply:
Good point! Committed your suggested fix.